builder/dockerfile/buildargs.go

```diff
@@ -42,6 +42,26 @@ func newBuildArgs(argsFromOptions map[string]*string) *buildArgs {
 	}
 }
 
+func (b *buildArgs) Clone() *buildArgs {
+	result := newBuildArgs(b.argsFromOptions)
+	for k, v := range b.allowedBuildArgs {
+		result.allowedBuildArgs[k] = v
+	}
+	for k, v := range b.allowedMetaArgs {
+		result.allowedMetaArgs[k] = v
+	}
+	for k := range b.referencedArgs {
+		result.referencedArgs[k] = struct{}{}
+	}
+	return result
+}
+
+func (b *buildArgs) MergeReferencedArgs(other *buildArgs) {
```

Comment:
Hmm, I find that more confusing than `MergeReferecedArgs`. The idea there is to merge the referenced args lists (each build stage has its own buildArgs instance now, so at the end, we merge the referenced args altogether to be able to warn if a given build arg was not consumed)